docs/library/network.rst

```diff
@@ -29,6 +29,95 @@ For example::
     data = s.recv(1000)
     s.close()
 
+Common network interface
+========================
+
+.. class:: BaseNIC(id=None, ...)
+
+Create a network interface object. Parameters are network interface
+dependent. If there are few interfaces of the same type, the first
+parameter should be `id`.
+
+    .. method:: active([is_active])
+
+        Activate ("up") or deactivate ("down") network interface, if boolean
+        argument is passed. Otherwise, query current state if no argument is
+        provided. Most other methods require an active interface.
+
+    .. method:: connect([service_id, key=None, \*, ...])
+
+       Connect interface to the network. This method is optional, and
+       available only for interfaces which are not "always connected".
+       If no parameters are given, connect to the default (or only)
+       service. If a single parameter is given, it is primary identifier
+       of a service to connect to. It may be accompanied by a key
+       (password) required to access service. There can be further
+       arbitrary keyword-only parameters, depending on networking medium
+       type and/or particular device. Parameters can be used to: a)
+       specify alternative service identifer types; b) provide additional
+       connection parameters. For various medium types, there are different
+       sets of predefined/recommended paramaters, among them:
+
+       * WiFi: `bssid` keyword to connect by BSSID (MAC address) instead
+         of access point name
+
+    .. method:: disconnect()
+
+       Disconnect from network.
+
+    .. method:: isconnected()
+
+       Returns True if connected to network.
+
+    .. method:: scan(\*, ...)
+
+       Scan for the available network services/connections. Returns
+       list of tuples with discovered service parameters. For various
+       network media, there are different variants of predefined/
+       recommended tuple formats, among them:
+
+       * WiFi: (ssid, bssid, channel, RSSI, authmode, hidden). There
+       may be further fields, specific to a particular device.
+
+       The may accept additional keyword arguments to filter scan results
+       (e.g. scan for a particular service, on a particular channel,
+       for services of a particular set, etc.), and to affect scan
+       duration and other parameters. Where possible, parameter names
+       should match those in connect().
+
+    .. method:: status()
+
+       Return detailed status of the interface, values are dependent
+       on the network medium/technology.
+
+    .. method:: ifconfig([(ip, subnet, gateway, dns)])
```

Comment:
Thanks for the feedback Paul.  I'll work on implementing the standard interface with our codebase, and a proposal for ways to extend it if I find myself stuck.

How about documenting that `ifconfig()` could return `None` as one of the IP addresses if it doesn't apply or isn't available?  Or would you expect `'0.0.0.0'` in that case?  If a user wants to set the DNS server without changing existing settings, I suppose they would do something like:

    c = nic.ifconfig()
    nic.ifconfig((c[0], c[1], c[2], '8.8.4.4'))  # change DNS without modifying other settings

I would propose documenting (and supporting) using `None` as placeholders in the tuple for elements that you don't want to change.

    nic.ifconfig((None, None, None, '8.8.4.4'))  # change DNS without modifying other settings